packages/enketo-express/public/js/src/module/controller-webform.js

```diff
@@ -443,9 +444,28 @@ function _submitRecord(survey) {
                     );
                 }, 1200);
             } else {
-                msg = msg.length > 0 ? msg : t('alert.submissionsuccess.msg');
-                gui.alert(msg, t('alert.submissionsuccess.heading'), level);
-                _resetForm(survey);
+                const parser = new DOMParser();
+                const xmlForm = parser.parseFromString(
+                    data.modelStr,
+                    'text/xml'
+                );
+                const maybeSubmitMessage =
+                    xmlForm.querySelector('submitMessage');
```

Comment:
> Kobo BE picks this up in the originating POST's submission XML

Great, so this looks like code that belongs in Kobo's backend, not in Enketo.

If it is indeed a requirement to also change the text displayed at the returnUrl, I see two options: 
* have your server serve a page to redirect to for that form. If a submit message is specified, it is displayed on that page and that page's URL is included as the return URL for the form
* read the response message from the OpenRosa response as with the no-redirect case. Change the default redirect page shipped with Enketo so that a message can be injected into it

Reply:
Note, some backends are already returning a `<message>` string in the `<OpenRosaResponse>` body, but it is being completely ignored and Enketo is displaying a boilerplate response to the user anyway. So in order to retain this existing desired (but arguably incorrect?) behavior of Enketo - ie not displaying the `<message>` in a lot of cases - the decision was made to first check whether to form just sent actually had a "kobo:submitMessage" setting in it, and only display the `<message>` body if it did [yes, this is a bit of a hack, but necessary to retain existing behavior of Enketo when talking to backends that return a `<message>` today, but it is being ignored...] But, again, this check should be for the presence of the _custom attribute_, not any hardcoded specific XML element .

Reply:
Ok, I better see the flow here, I was trying to go too quickly. Now I see this is where the message is alerted and has nothing to do with the return url. 

> yes, this is a bit of a hack

Indeed. At the very least I think it should check for the `kobo` namespace. Right now, unless I've misunderstood something else, the first attribute or element with name `submitMessage` in the default namespace will be a match. You could look at some of the OpenClinica-specific behavior for examples of introducing and isolating behavior for a specific platform.

Reply:
So this 'hack' is undesirable, and is only necessary to - ostensibly - persist Enketo's (arguably) incorrect current behavior of completely ignoring the `<message>` being returned in the OpenRosaResponse, and instead displaying a hardcoded message instead. Ideally, a boilerplate message would _only_ be displayed if the message were, say, empty (if even then), whereas any `<message>` actually provided would be displayed, as intended! That would seem to be entirely the intent of having this optional message element in the OpenRosaResponse body.

So my preference would be to remove this hack, and have Enketo always display - for the post-submission popup - whatever `<message>` in received. And to only display the existing generic boilerplate message when no message is provided.

The potential impact of doing this (ie removing this hack, for checking for the special attribute in the original form...) is that Enketo would now actually start displaying whenever OpenRosaResponse `<message>`s that existing backends might already happen to be returning (!), but which previously Enketo has been hiding. I dont think this is necessarily a 'bad' thing, per se, and if this is undersirable then these backends can always start returning an empty `<message>` instead to retain Enketo's existing boilerplate followup response.

Again, the explicit intent of the `<message>` in the OpenRosaResponse is to provide a mechanism for a BE to return a customized post-submission message to the client upon successful submission. Changing Enketo to allow this to include markdown seems like an obvious, consistent, and pretty natural way to accomplish this desired feature.

Reply:
> Enketo's (arguably) incorrect current behavior of completely ignoring the <message> being returned in the OpenRosaResponse

I agree that the original spec intent was for the message to come from the server. I think both Collect and Enketo ended up taking control of success messages and some standard error messages because it means they can provide a more uniform, coherent user experience and take control of localization.

I'm not immediately against changing that behavior but I think it warrants its own broader discussion pulling in other server implementers. I think we'd also want to have a plan for localization. That would entail having Enketo send `Accept-Language` headers (https://docs.getodk.org/openrosa-http/#http-requests-and-responses). Servers would then be expected to have localized default messages to return. I think it's a fair amount of work across a lot of code bases.

A somewhat less hacky option would be to document a special message `nature` (https://docs.getodk.org/openrosa-http/#http-responses) that clients could recognize and use to know they should override their default strings. For example, it could have a value like `__client_default_override`. Servers could specify messages of this nature when they know they have something more useful than the client default to say.

Reply:
Oh, and I don't know whether this is what @boazsender might be referring to with "moving this work to the single route" but you could also entirely bypass OpenRosa messages by targeting the path with `returnUrl` instead.

Reply:
I was confused when I said I was moving this work to the `single` route. Please disregard that comment.